File: backend/lib/Analysis.jl
module Analysis

using HTTP, JSON, DataFrames, Survival, StatsModels, StatsBase, CBioPortal

function parse_config(jsonstring)
    return JSON.Parser.parse(jsonstring)
end

function load_clinical_subset(config, clinical_data, survival_meta)
    study_id = config["study_id"]
    subset = clinical_data[(clinical_data.STUDY_ID.==study_id), :]
    meta = survival_meta[survival_meta.ID.==study_id, ["SURVIVAL_OUTCOME", "SURVIVAL_TIME_UNIT"]]
    outcome = meta[1][1]
    time_unit = meta[2][1]
    return (subset, outcome, time_unit)
end


function preprocess_moleculardata(response, config)
    # Initialise df
    df = DataFrame(PATIENT_ID=String[])
    for gene in config["genes"]
        df[:, string(gene["entrez"])] = Union{Float64, Missing}[]
    end

    # Fill temp dict
    dict = Dict()
    for object in response
        patient_id = object["patientId"]
        value = object["value"]
        entrez = object["entrezGeneId"]
        if haskey(dict, patient_id)
            dict[patient_id][entrez] = value
        else
            dict[patient_id] = Dict(entrez => value)
        end
    end

    # Fill df
    for (patient, values) in dict
        new_patient::Array{Any} = [ patient ]
        for gene in config["genes"]
            value = get(values, gene["entrez"], missing)
            push!(new_patient, value)
        end
        push!(df, new_patient)
    end

    dropmissing!(df)

    # Group based on columns
    groups = []
    percentiles = Dict(string(g["entrez"]) => StatsBase.percentile(df[:, string(g["entrez"])], g["percentile"]) for g in config["genes"])
    opposites = Dict(string(g["entrez"]) => StatsBase.percentile(df[:, string(g["entrez"])], 100-g["percentile"]) for g in config["genes"])
    # mirrored controls fulfil all the criteria of the test in the opposite percentile
    # complement controls are any cases not in the test group
    # NOTE that these are set per-gene
    # Also note that a control must be a control in ALL genes to be classified as a control
    # eg. if a case fulfils one gene criteria ("test group") but not another gene criteria ("control group") it will be completely excluded (neither group)
    for row in eachrow(df)
        satisfied_test = []
        satisfied_control = []
        for gene in config["genes"]
            entrez = string(gene["entrez"])
            percentile = percentiles[entrez]
            opposite = opposites[entrez]
            if gene["side"] == "above"
                single_satisfy_test = row[entrez] >= percentile
                if gene["control_type"] == "mirrored"
                    single_satisfy_control = row[entrez] < opposite
                elseif gene["control_type"] == "complement"
                    single_satisfy_control = !(single_satisfy_test)
                end
            elseif gene["side"] == "below"
                single_satisfy_test = row[entrez] <= percentile
                if gene["control_type"] == "mirrored"
                    single_satisfy_control = row[entrez] > opposite
                elseif gene["control_type"] == "complement"
                    single_satisfy_control = !(single_satisfy_test)
                end
            end
            push!(satisfied_test, single_satisfy_test)
            push!(satisfied_control, single_satisfy_control)
        end
        if all(satisfied_test)
            push!(groups, "TEST")
        elseif all(satisfied_control)
            push!(groups, "CONTROL")
        else
            push!(groups, "NEITHER")
        end
    end
    df.GROUP = groups
    return df
end


function join_data(clinicaldata, moleculardata)
    joined = innerjoin(clinicaldata, moleculardata, on=:PATIENT_ID)
    return joined
end

function preprocess_km(km_fit)
    # Steps
    steps = Dict(
        "times" => km_fit.times,
        "survival" => km_fit.survival
    )

    # Scatter
    censor_times = []
    censor_survival = []
    for (i, censored) in enumerate(km_fit.ncensor)
        if censored > 0
            push!(censor_times, km_fit.times[i])
            push!(censor_survival, km_fit.survival[i])
        end
    end
    censors = Dict(
        "times" => censor_times,
        "survival" => censor_survival
    )

    # Aggregating
    result = Dict(
        "steps" => steps,
        "censors" => censors
    )

    return result
end

function perform_km(joined)
    controls = joined[joined.GROUP .== "CONTROL", :]
    test = joined[joined.GROUP .== "TEST", :]

    km_control = fit(KaplanMeier, controls.TIME, controls.STATUS)
    km_test = fit(KaplanMeier, test.TIME, test.STATUS)

    km_results = Dict(
        "control" => preprocess_km(km_control),
        "test" => preprocess_km(km_test)
        )
    return km_results
end

function perform_cox(joined)
    subset = filter(row -> (row.GROUP == "CONTROL" || row.GROUP=="TEST"), joined)
    subset.EVENT = EventTime.(subset.TIME, Bool.(subset.STATUS))
    subset.GROUP_INT = map(row -> row.GROUP == "CONTROL" ? 0 : row.GROUP == "TEST" ? 1 : -1, eachrow(subset))
    cox_model = coxph(@formula(EVENT ~ GROUP_INT), subset)
    coefs = coeftable(cox_model)
    results = Dict(
        "p_value" => coefs.cols[4][1],
        "estimate" => coefs.cols[1][1],
        "hr" => exp(coefs.cols[1][1]),
    )
    return results
end

function perform_survival_analysis(config, all_clinical_data, all_survival_meta)
    local clinicaldata, clinicalevent, clinicaltimeunit, moleculardata
    @sync begin
        @async begin
            moleculardata_dict = CBioPortal.get_moleculardata(config)
            moleculardata = preprocess_moleculardata(moleculardata_dict, config)
        end
        @async begin
            (clinicaldata, clinicalevent, clinicaltimeunit) = load_clinical_subset(config, all_clinical_data, all_survival_meta)
        end
    end

    # Join the data
    joined = join_data(clinicaldata, moleculardata)
    total_num_joined = size(joined)[1]
    num_control = sum(joined.GROUP.=="CONTROL")
    num_test = sum(joined.GROUP.=="TEST")
    total_num_analysis = num_control + num_test

    # Perform statistics
    km = perform_km(joined)
    cox = perform_cox(joined)

    # Aggregate
    result = Dict(
        "km" => km,
        "cox" => cox,
        "event_type" => clinicalevent,
        "event_time_unit" => clinicaltimeunit,
        "num_available" => total_num_joined,
        "num_analysis" => total_num_analysis,
        "num_control" => num_control,
        "num_test" => num_test,
        "config" => config,
    )

    return result
end


end

File: backend/src/analysis.py
"""Functions for main survival analysis.

The main survival analysis looks at the combined effect of one or more genes
on cancer survival outcomes. Genes are combined based on percentile
thresholds within a cancer study population.

The key steps to performing the analysis are:

1. Extracting clinical outcomes from the study population
2. Extracting genetic expression for selected genes from the study population 
3. Binning the genetic expression data into a test and control group based on percentile thresholds
4. Joining the binned genetic expression data and clinical outcomes 
5. Correlating outcomes ~ group through survival analysis

"""

import pandas as pd
import numpy as np
import src.cbioportal as cbioportal
from typing import List
from lifelines import CoxPHFitter, KaplanMeierFitter


def allocate_group(column, threshold: float, direction: str, control: str):
    """Allocates a column's percentile values into a test or control group (or neither).

    Args:
        column: column of PERCENTILE data
        threshold (float): float threshold.
        direction (str): one of "above" or "below".
        control (str): one of "mirrored" or "complement".

    Returns:
        New column of values, one of:
        - 1 if satisfies test condition.
        - 0 if satisfies control condition.
        - -1 if satisfies neither.
    """
    if control == "complement":
        if direction == "above":
            return np.where(column > threshold, 1, 0)
        elif direction == "below":
            return np.where(column < threshold, 1, 0)
        else:
            raise Exception(
                f"Unknown threshold direction: {direction}. 'above' or 'below' expected."
            )
    elif control == "mirrored":
        if direction == "above":
            tests = np.where(column > threshold, 1, -1)
            controls = np.where(column < (1 - threshold), 1, 0)
            return tests + controls
        elif direction == "below":
            tests = np.where(column < threshold, 1, -1)
            controls = np.where(column > (1 - threshold), 1, 0)
            return tests + controls
    else:
        raise Exception(
            f"Unknown control type: {control}. 'mirrored' or 'complement' expected."
        )


def apply_thresholds(expression: pd.DataFrame, thresholds: List) -> pd.Series:
    """Allocates each row into test (1) and control (0) or neither (NaN).

    Args:
        expression (pd.DataFrame): mrna expression data
        thresholds (float): thresholds to apply. This should be a list of dicts
            where each dict has the following keys:
            - gene which is a dict with entrez (int): the entrez ID of the gene
            - threshold (float) the threshold (test does not include threshold
              value).
            - direction (str): one of "above" or "below" specifying direction
              of test group relative to threshold
            - control (str): one of "mirrored" or "complement" specifying how
              controls are defined. Mirrored controls are on the opposite side
              of (1 - threshold). Complement controls are on the opposite side
              of threshold.

    Returns:
        A column with values equal to 0 (control), 1 (test) or NaN.
    """

    def combine_groups(row):
        """Combine threshold groups in each row.
        - If all 0s, then 0
        - If all 1s, then 1
        - Else NaN
        """
        if np.all(row == 1):
            return 1
        elif np.all(row == 0):
            return 0
        else:
            return np.nan

    column_groups = []
    percentiles = expression.rank(pct=True)
    for thresh in thresholds:
        group = allocate_group(
            percentiles[thresh["gene"]["entrez"]],
            thresh["threshold"],
            thresh["direction"],
            thresh["control"],
        )
        column_groups.append(group)
    column_groups = pd.DataFrame(np.stack(column_groups, axis=1))
    combined = column_groups.apply(combine_groups, axis=1)
    return combined


def group_and_join(clinical, molecular, thresholds, outcome) -> pd.DataFrame:
    """Group the molecular data based on provided thresholds and join into a single dataframe without NaNs.

    Any rows which contain NaN are dropped.
    """
    molecular["group"] = apply_thresholds(molecular, thresholds).values
    joined = clinical.join(molecular, how="inner")
    if outcome == "os":  # remove irrelevant columns
        columns = [c for c in joined.columns if c not in ["dfs_months", "dfs_status"]]
        subset = joined[columns].copy()
        renamed = subset.rename(columns={"os_months": "months", "os_status": "status"})
    elif outcome == "dfs":
        columns = [c for c in joined.columns if c not in ["os_months", "os_status"]]
        subset = joined[columns].copy()
        renamed = subset.rename(
            columns={"dfs_months": "months", "dfs_status": "status"}
        )
    renamed.dropna(inplace=True)
    renamed["months"] = renamed["months"].astype(float)
    renamed["status"] = renamed["status"].astype(float).astype(int)
    renamed["group"] = renamed["group"].astype(int)
    return renamed


def perform_cph(data, outcome):
    """Perform Cox proportional hazards analysis given the clinical and molecular data and thresholds.

    Args:
        data (pd.DataFrame): dataframe with at least columns "months", "status" and "group".
        outcome (string): one of "dfs" or "os".
    """
    subset = data[["months", "status", "group"]]
    cph = CoxPHFitter()
    cph.fit(subset, "months", "status")
    return cph


def perform(api: cbioportal.Api, config):
    """Perform an analysis for a given configuration.

    The configuration is a dict which must contain:
    - study_id (str): study ID
    - profile_id (str): molecular profile ID
    - outcome_id (str): one of "os" or "dfs"
    - thresholds: list of threshold dicts

    Example configuration:

        {
            "analysis_id": 0,
            "study_id": "brca_metabric",
            "profile_id": "brca_metabric_mrna",
            "outcome_id": "os",
            "thresholds": [
                {
                    "gene": { "entrez": 675 },
                    "threshold": 0.7,
                    "direction": "above",
                    "control": "complement",
                },
                {
                    "gene": { "entrez": 5728 },
                    "threshold": 0.3,
                    "direction": "below",
                    "control": "mirrored",
                },
            ]
        }

    Returns either:
        - (None, { "type": "api" | , "message": string, "debug": string })
        - (result, False)

    Example result:


    """
    outcome = config["outcome_id"]
    thresholds = config["thresholds"]
    entrez = [t["gene"]["entrez"] for t in thresholds]

    try:
        clinical = api.get_clinical_data(config["study_id"])
        molecular = api.get_molecular_data(entrez, config["profile_id"])
    except Exception as e:
        return None, {
            "type": "api",
            "message": "There was an API error. Let us know.",
            "debug": str(e),
        }

    joined = group_and_join(clinical, molecular, thresholds, outcome)

    if len(joined) < 20:
        return None, {
            "type": "data",
            "message": f"There was insufficient data to perform the analysis (n={len(joined)} after joining clinical data with available molecular data).",
            "debug": "",
        }

    try:
        cph = perform_cph(joined, outcome)
    except Exception as e:
        return None, {
            "type": "model",
            "message": "Performing the Cox Proportional Hazards analysis failed during fitting.",
            "debug": str(e),
        }

    num_test = int((joined.group == 1).sum())
    num_control = int((joined.group == 0).sum())
    num_clinical = int(clinical.count().max())
    num_excluded = num_clinical - (num_test + num_control)

    csv_data = joined.to_csv()

    def is_right_censor(row, df):
        "Determines if a row is right censored i.e. outcome is 0."
        return 0 in df[df.months == row["timeline"]].status.values

    test_kmf = KaplanMeierFitter()
    test_joined = joined[joined.group == 1]
    test_kmf.fit(test_joined.months, test_joined.status)
    test_km_data = (
        test_kmf.survival_function_.reset_index().astype("float").to_dict("records")
    )
    test_km_censors = [r for r in test_km_data if is_right_censor(r, test_joined)]

    cont_kmf = KaplanMeierFitter()
    cont_joined = joined[joined.group == 0]
    cont_kmf.fit(cont_joined.months, cont_joined.status)
    cont_km_data = (
        cont_kmf.survival_function_.reset_index().astype("float").to_dict("records")
    )
    cont_km_censors = [r for r in cont_km_data if is_right_censor(r, cont_joined)]

    hr = float(cph.hazard_ratios_.group.item())
    p = float(cph.summary.p.group.item())

    result = {
        "analysis_id": config["analysis_id"],
        "num_test": num_test,
        "num_control": num_control,
        "num_clinical": num_clinical,
        "num_excluded": num_excluded,
        "outcome": outcome,
        "outcome_units": "months",  # same for both os and dfs
        "hazard_ratio": hr,
        "p_value": p,
        "test_km_data": test_km_data,
        "test_km_censors": test_km_censors,
        "cont_km_data": cont_km_data,
        "cont_km_censors": cont_km_censors,
        "csv_data": csv_data,
    }
    return result, False
